src/OpenSolid/Sphere3d.elm

```diff
@@ -0,0 +1,500 @@
+            { centerPoint =
+                Point3d.fromCoordinates ( 1, 2, 1 )
+            , radius = 3
+            }
+
+-}
+with : { centerPoint : Point3d, radius : Float } -> Sphere3d
+with properties =
+    Internal.Sphere3d { properties | radius = abs properties.radius }
+
+
+{-| Construct a sphere around the given axis that passes through the given point.
+
+    point : Point3d
+    point =
+        Point3d.fromCoordinates ( 3, 0, 2 )
+
+    Sphere3d.around Axis3d.z point
+    --> Sphere3d.with
+    -->     { centerPoint =
+    -->         Point3d.fromCoordinates ( 0, 0, 2 )
+    -->     , radius = 3
+    -->     }
+
+-}
+around : Axis3d -> Point3d -> Sphere3d
+around axis point =
+    let
+        centerPoint =
+            Point3d.projectOntoAxis axis point
+    in
+    with
+        { centerPoint = centerPoint
+        , radius = Point3d.distanceFrom centerPoint point
+        }
+
+
+{-| Attempt to construct a sphere that passes through the four given points.
+Returns `Nothing` if the first three of the four given points are collinear.
+Also returns `Nothing` if the fourth point lies in a plane with the three other ones.
+
+    Sphere3d.throughPoints
+        ( Point3d.fromCoordinates (1, 0, 0)
+        , Point3d.fromCoordinates (-1, 0, 0)
+        , Point3d.fromCoordinates (0, 1, 0)
+        , Point3d.fromCoordinates (0, 0, 0.5)
+        )
+    --> Just <| Sphere3d.with
+        { centerPoint =
+            Point3d.fromCoordinates (0, 0, -0.75)
+            , radius = 1.25
+            }
+
+    Sphere3d.throughPoints
+        ( Point3d.fromCoordinates (1, 0, 0)
+        , Point3d.fromCoordinates (-1, 0, 0)
+        , Point3d.fromCoordinates (0, 1, 0)
+        , Point3d.fromCoordinates (0, -1, 0)
+        )
+    --> Nothing
+
+-}
+throughPoints : ( Point3d, Point3d, Point3d, Point3d ) -> Maybe Sphere3d
+throughPoints ( p1, p2, p3, p4 ) =
+    Maybe.andThen
+        {-
+           First three points define a circle.
+           All points M on the normal to this circle through the circle's center are equidistant to p1, p2 and p3.
+           Therefore: find this point M such that it is equidistant to p1, p2, p3 and p4,
+           this will be the center of the sphere.
+        -}
+        (\circle ->
+            let
+                normalAxis =
+                    Circle3d.axis circle
+
+                r =
+                    Circle3d.radius circle
+
+                x =
+                    Point3d.distanceFromAxis normalAxis p4
+
+                y =
+                    Point3d.distanceAlong normalAxis p4
+
+                d =
+                    (r ^ 2 - x ^ 2 - y ^ 2) / (-2 * y)
+            in
+            if y /= 0 then
+                Just <|
+                    with
+                        { centerPoint = Point3d.along normalAxis d
+                        , radius = sqrt (r ^ 2 + d ^ 2)
+                        }
+            else
+                Nothing
+        )
+        (Circle3d.throughPoints ( p1, p2, p3 ))
```

Comment:
Minor point but I think piping would make this a bit more readable:

```elm
Circle3d.throughPoints ( p1, p2, p3 )
    |> Maybe.andThen
        (\circle ->
            ...
        )
```

Reply:
👍